### DISCUSSION TITLE
How to setup custom SMTP from config.toml for local development setup?

### DISCUSSION THREAD

**Discussion by atulrpandey** [supabase/supabase#27054] [Questions]
I am using Supabase local setup and configuring email auth to send a confirmation email upon successful sign-up. How do I configure a custom SMTP in my local setup to send the emails? What configuration fields in config.toml control this?

**Comment by GaryAustin1** [comment]
Local dev uses an "in bucket"port for showing what would be sent in email. 

If you haven't changed the setting it should be on http://localhost:54324/ I believe.

**Reply by atulrpandey** [reply]
Thanks, @GaryAustin1. Local dev with "in bucket" is working as expected. I wanted to control the configuration for custom SMTP from the **config.toml** file.  Can the configuration for custom SMTP be controlled from the **config.toml** file?

**Reply by GaryAustin1** [reply]
Yes.   https://github.com/orgs/supabase/discussions/2818#discussioncomment-4019776

**Reply by atulrpandey** [reply]
@GaryAustin1, the discussion link you shared concerns the social login providers. If I am missing something, please direct me to the correct comment, where the custom SMTP setup (with Supabase CLI) for sending emails is discussed.

I also found this discussion - https://github.com/orgs/supabase/discussions/22677 which says it is impossible. If there is an update to this, please let me know. Thanks.

**Reply by GaryAustin1** [reply]
Oh sorry.  Early morning brain blip.   
You can't set SMTP on local dev.  The in bucket is it.

**Reply by atulrpandey** [reply]
No issues! I can understand. 😅 

This means that after we deploy the Supabase app using the CLI, we will have to configure SMTP manually on the dashboard.

**Comment by Clactar** [comment]
If we can't use custom STMP on CLI, then why is there a [auth.email.smtp] part on the config.toml ?
Does it mean that the supabase CLI can be used in prod?

**Reply by GaryAustin1** [reply]
For selfhosted or Supabase hosted instances you can.  Just not for local dev.

**Reply by PatoSala** [reply]
So if I set up a custom smtp using config.toml, will local dev still use mailpit? And if I push config.toml to remote will remote start using the smtp that I previously set in config.toml?

**Comment by deepshekhardas** [comment]
To clarify both the original question and @PatoSala's follow-up:

**Local dev always uses Mailpit (inbucket) — regardless of what you set in `config.toml`.**

The `[auth.email.smtp]` section in `config.toml` is intentionally ignored by the Supabase CLI for local development. This is by design — the CLI routes all outgoing emails to Mailpit at `http://localhost:54324/` so you can inspect them without actually sending anything.

**To answer @PatoSala directly:**
- Yes, even if you define custom SMTP in `config.toml`, local dev will still use Mailpit.
- When you push/deploy to a remote Supabase-hosted project, the SMTP settings from `config.toml` **will** be applied there.
- So the config.toml SMTP block is only relevant for **self-hosted** or **Supabase-hosted remote** environments.

**Practical workflow:**
```
Local dev → Mailpit (localhost:54324) — no config needed
Remote/self-hosted → [auth.email.smtp] in config.toml applies
```

If you need to test real SMTP delivery locally, one workaround is to use a tool like [Mailtrap](https://mailtrap.io) or [Resend](https://resend.com) with a test API key and configure it in your remote Supabase project's Auth settings directly via the dashboard.

**Reply by PatoSala** [reply]
> To clarify both the original question and @PatoSala's follow-up:
> 
> 
> 
> **Local dev always uses Mailpit (inbucket) — regardless of what you set in `config.toml`.**
> 
> 
> 
> The `[auth.email.smtp]` section in `config.toml` is intentionally ignored by the Supabase CLI for local development. This is by design — the CLI routes all outgoing emails to Mailpit at `http://localhost:54324/` so you can inspect them without actually sending anything.
> 
> 
> 
> **To answer @PatoSala directly:**
> 
> - Yes, even if you define custom SMTP in `config.toml`, local dev will still use Mailpit.
> 
> - When you push/deploy to a remote Supabase-hosted project, the SMTP settings from `config.toml` **will** be applied there.
> 
> - So the config.toml SMTP block is only relevant for **self-hosted** or **Supabase-hosted remote** environments.
> 
> 
> 
> **Practical workflow:**
> 
> ```
> 
> Local dev → Mailpit (localhost:54324) — no config needed
> 
> Remote/self-hosted → [auth.email.smtp] in config.toml applies
> 
> ```
> 
> 
> 
> If you need to test real SMTP delivery locally, one workaround is to use a tool like [Mailtrap](https://mailtrap.io) or [Resend](https://resend.com) with a test API key and configure it in your remote Supabase project's Auth settings directly via the dashboard.

Awesome! Thanks for the clear explanation! I am currently trying to set up custom smtp and I'd like to not use my remote supabase dashboard because I have a github action to push config.toml changes to remote. And at the same time I like usong mailpit for development, its way more efficient than having to use real emails for testing.

**Reply by PatoSala** [reply]
Hey I wanted to give an update about custom smtp in local supabase instance. Supabase local instance does use custom smtp if its set on config.toml, its not like it ignores the config and uses Mailpit instead. I just find out after setting my custom smtp.

**Reply by felipebn** [reply]
+1 on the above.

I only managed to get emails into mailpit after commenting out the section entirely.
